pkg/util/containers/v2/metrics/collector_containerd.go

```diff
@@ -0,0 +1,416 @@
+	cutil "github.com/DataDog/datadog-agent/pkg/util/containerd"
+	"github.com/DataDog/datadog-agent/pkg/util/containers/providers"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+	"github.com/DataDog/datadog-agent/pkg/util/system"
+)
+
+const (
+	containerdCollectorID    = "containerd"
+	containerdCacheKeyPrefix = "containerd-stats-"
+	containerdCacheTTL       = 10 * time.Second
+)
+
+func init() {
+	metricsProvider.registerCollector(collectorMetadata{
+		id:       containerdCollectorID,
+		priority: 1, // Less than the "system" collector, so we can rely on cgroups directly if possible
+		runtimes: []string{RuntimeNameContainerd},
+		factory: func() (Collector, error) {
+			return newContainerdCollector()
+		},
+	})
+}
+
+type containerdCollector struct {
+	client         cutil.ContainerdItf
+	lastScrapeTime time.Time
+}
+
+func newContainerdCollector() (*containerdCollector, error) {
+	client, err := cutil.GetContainerdUtil()
+	if err != nil {
+		return nil, ErrNothingYet
+	}
+
+	return &containerdCollector{client: client}, nil
+}
+
+func (c *containerdCollector) ID() string {
+	return containerdCollectorID
+}
+
+func (c *containerdCollector) GetContainerStats(containerID string, cacheValidity time.Duration) (*ContainerStats, error) {
+	refreshRequired := c.lastScrapeTime.Add(cacheValidity).Before(time.Now())
+	cacheKey := containerdCacheKeyPrefix + containerID
+	if cacheStats, found := cache.Cache.Get(cacheKey); found && !refreshRequired {
+		stats := cacheStats.(*ContainerStats)
+		log.Debugf("Got containerd stats from cache for container %s", containerID)
+		return stats, nil
+	}
+
+	metrics, err := c.getMetrics(containerID)
+	if err != nil {
+		return nil, err
+	}
+
+	c.lastScrapeTime = time.Now()
+
+	var res *ContainerStats
+
+	switch metricsVal := metrics.(type) {
+	case *v1.Metrics:
+		container, err := c.client.Container(containerID)
+		if err != nil {
+			return nil, fmt.Errorf("could not get container with ID %s: %s", containerID, err)
+		}
+
+		OCISpec, err := c.client.Spec(container)
+		if err != nil {
+			return nil, fmt.Errorf("could not retrieve OCI Spec from container with ID %s: %s", containerID, err)
+		}
+
+		info, err := c.client.Info(container)
+		if err != nil {
+			return nil, fmt.Errorf("could not retrieve the metadata of the container with ID %s: %s", containerID, err)
+		}
+
+		processes, err := c.client.TaskPids(container)
+		if err != nil {
+			return nil, fmt.Errorf("could not retrieve the processes of the container with ID %s: %s", containerID, err)
+		}
+
+		res = getContainerdStatsLinux(metricsVal, info, OCISpec, processes)
+	case *wstats.Statistics:
+		windowsMetrics := metricsVal.GetWindows()
+
+		if windowsMetrics == nil {
+			return nil, fmt.Errorf("error getting Windows metrics for container with ID: %s: %s", containerID, err)
+		}
+
+		res = getContainerdStatsWindows(windowsMetrics)
+	default:
+		return nil, fmt.Errorf("can't convert the metrics data from container with ID %s", containerID)
```

Comment:
Can we print the type we got using `%T` ?

Reply:
Good idea. I've changed a couple of error messages to include the type :+1: